feat(eslint-plugin): new rule prefer-signals

reduckted opened the pull request:
Fixes #1719 and #1824.

This rule checks a few things.

### Prefer Signals

This rule will report a problem when a `BehaviorSubject` is created. This type name can be overridden in the options. That allows this part of the rule to be skipped (by specifying an empty array), or to include additional types that should be replaced with signals.

There is no suggestion for this part of the rule.

### Prefer Readonly

This rule will report a problem for any class property that does not have the `readonly` modifier and either:
* Has a type annotation of:
    * `Signal`
    * `InputSignal`
    * `ModelSignal`
    * `WritableSignal`
* Is assigned the result of:
    * `computed()`
    * `contentChild()`
    * `contentChild.required()`
    * `contentChildren()`
    * `input()`
    * `input.required()`
    * `model()`
    * `signal()`
    * `viewChild()`
    * `viewChild.requried()`
    * `viewChildren()`
* If type checking is enabled, the type is computed to be a type of signal (the same type names that the type annotation is checked for).

This part of the rule can be disabled.

### Prefer Input Signals

This rule will report a problem when `@Input` is used. This part of the rule can be disabled.

### Prefer Query Signals

This rule will report a problem when `@ContentChild`, `@ContentChildren`, `@ViewChild` or `@ViewChildren` is used. This part of the rule can be disabled.

reduckted commented:
[`toSignal`](https://angular.dev/api/core/rxjs-interop/toSignal) is another function to check for.

MillerSvt commented on packages/eslint-plugin/src/rules/prefer-signal-readonly.ts:
```diff
@@ -0,0 +1,105 @@
+  'Signal',
+  'WritableSignal',
+]);
+const KNOWN_SIGNAL_CREATION_FUNCTIONS: ReadonlySet<string> = new Set([
```
I think we should check the type of the signal instead of the function name. There are many custom utility functions that return signals.

P4 replied:
Type-aware linting used to have a significant performance cost, is this still the case in newer versions? If so, having a faster but more limited rule may be more desirable for some people.

One way to solve this would probably be to make using types opt-in, but I got no idea if that's something that can be done inside a single rule.

MillerSvt replied:
A lot of people already have type-aware linting, so I don't think that one more rule will greatly affect performance.

Adding an option to disable it is a good idea, I'll give it a try.

MillerSvt replied:
https://github.com/reduckted/angular-eslint/pull/1

JamesHenry commented:
I'm strongly thinking we will want to have a `prefer-signals` rule very soon, and having `prefer-signals` and `prefer-signal-readonly` is maybe a little clunky...

It feels like `readonly` is pretty objectively desirable for these APIs, so maybe we make it so that `prefer-signals` also enforces `readonly` by default, but that rule has an option to disable it if people strongly dislike it for some reason? 🤔 

I do agree with @MillerSvt that we need to at least have the option to use type information for enforcing signals usage.

@reduckted would you like to collaborate on the `prefer-signals` implementation?

reduckted commented:
Makes sense to just have a single `prefer-signals` rule with a bunch of options, so I'm happy for this to not be merged.

I'd be happy to help out with that rule. Is #1824 the existing issue about it?

In the meantime, I'll switch this PR to draft and update it to include type checking just so we have the implementation available somewhere.

reduckted commented:
I've added two options:
* `useTypeChecking`: When true, type-checking will be used as a last resort.
* `signalCreationFunctions`: An array of user-defined function names that create signals. This allows those custom functions to be declared and treated like the built-in functions (`signal`, `model`, etc.) while also avoiding type-checking.

reduckted commented:
I've changed this rule to `prefer-signals` and added a few more options. I've updated my initial comment with everything that the rule now does.

Feel free to critique anything that doesn't seem right, or isn't implemented in the preferred style. 😁

pumano commented:
Would love to use it, when it merged.

JamesHenry commented on package.json:
```diff
@@ -127,5 +127,6 @@
     "includedScripts": []
-  }
+  },
+  "packageManager": "yarn@1.22.21+sha1.1959a18351b811cdeedbd484a8f86c3cc3bbaf72"
```
Please revert this, it shouldn't have been added in the first place, but I have also migrated from yarn classic to pnpm in the meantime and this now prevents CI from running

JamesHenry replied:
This was marked as resolved before it was addressed?

reduckted replied:
Yeah, I was marking things off as I made the changes locally. I've just pushed up all of the changes.

JamesHenry commented on packages/eslint-plugin/docs/rules/prefer-signal-readonly.md:
This file should be removed

JamesHenry replied:
This was marked as resolved before it was addressed?

reduckted commented on packages/eslint-plugin/tests/rules/prefer-signals/spec.ts:
```diff
@@ -0,0 +1,17 @@
+import rule, { RULE_NAME } from '../../../src/rules/prefer-signals';
+import { invalid, valid } from './cases';
+
+const ruleTester = new RuleTester({
```
Success! 🥳 

> I have a feeling that might be obscuring the underlying problem though.

I was right about that. I just had to look further up the test output. The error I posted was from the invalid test cases. The error from the valid tests cases was:

```
    Message:
      Should have no errors but had 1: [
      {
        ruleId: null,
        severity: 1,
        message: 'No matching configuration found for __placeholder__.js.',
        line: 0,
        column: 0,
        nodeType: null
      }
    ]
```

That led me to discover that `RuleTester` only set the filename for tests when `languageOptions.parserOptions.project` is set. I just had to change that to set the filename if `project` or `projectService` is set, and the tests started passing.

JanProchy reviewed (approved):
Looking great!

Mr-Jami commented:
Exactly what I am looking for. Looking forward to try it out when merged.

pumano commented:
any news on this?

JamesHenry commented:
I'm going to merge this into #2109 and apply some tweaks, many thanks again!

JamesHenry commented:
@reduckted I wasn't fully happy with the scope of the arbitrary typesToReplace and the implementation only considered things being instantiated with `new` which made sense for BehaviorSubject, but not the arbitrary user provided list. I have therefore removed that for the initial version, as well as tweaked some messages and options, but thank you again for tackling the lion's share of this

JamesHenry commented:
We can revisit preferring signals over arbitrary things in a dedicated follow up

Frotty commented:
Great addition, but the rule cannot be used, see https://github.com/angular-eslint/angular-eslint/issues/2149
